### PR TITLE
Ind vid

### PR DESCRIPTION
**Pull request by AkashSadda**

### CODE REVIEW

File: NumberValidator/Validators/IValidator.cs [hunk 1]
```diff
@@ -1,8 +1,8 @@
 ﻿namespace NumberValidator.Validators
 {
-    public interface IValidator
+    public interface IAadhaarValidator
```
Comment (pradeepgururani) [inline NumberValidator/Validators/IValidator.cs]: Why are you renaming the interface? Do you not know what an interface is for?

File: NumberValidator/Validators/IN/vid.cs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+{
+    public class InvalidLengthException : Exception { }
+    public class InvalidFormatException : Exception { }
+    public class InvalidChecksumException : Exception { }
+
+    public static class Vid
+    {
+        private static readonly Regex VidRegex = new Regex(@"^[2-9][0-9]{15}$");
+
+        public static string Compact(string number) =>
```
Comment (pradeepgururani) [inline NumberValidator/Validators/IN/vid.cs]: why new method implementation? what functionality is not provided by `Clear` method?

File: NumberValidator/Validators/IN/vid.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            var compactedNumber = Compact(number);
+            return $"XXXX XXXX XXXX {compactedNumber.Substring(12)}";
+        }
+
+        private static bool IsPalindrome(string number) =>
+            number == new string(number.Reverse().ToArray());
+
+        public static string GenerateValidVid()
+        {
+            string vid;
```
Comment (pradeepgururani) [inline NumberValidator/Validators/IN/vid.cs]: we use `var` and not assign explicit type.

File: NumberValidator/Validators/IN/vid.cs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            do
+            {
+                vid = GenerateRandomVid();
+            } while (IsPalindrome(vid) || !IsValid(vid));
+            return vid;
+        }
+
+        private static string GenerateRandomVid()
+        {
+            var rand = new Random();
```
Comment (pradeepgururani) [inline NumberValidator/Validators/IN/vid.cs]: name it properly. what are we saving by typing two letter less?

File: NumberValidator/Validators/IN/vid.cs [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            if (CalculateChecksum(number) != 0)
+            {
+                throw new InvalidChecksumException();
+            }
+        }
+
+        public static int CalculateChecksum(string number)
+        {
+            var reversedDigits = number.Reverse().Select(digit => int.Parse(digit.ToString())).ToArray();
+            int c = 0;
```
Comment (pradeepgururani) [inline NumberValidator/Validators/IN/vid.cs]: use `var`. you are very inconsistent. lots of places you are using `var` and then at some places you start using explicit types.

File: NumberValidator/Validators/IN/vid.cs [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+            for (int i = 0; i < reversedDigits.Length; i++)
```
Comment (pradeepgururani) [inline NumberValidator/Validators/IN/vid.cs]: use `var`.

File: NumberValidator/Validators/IN/vid.cs [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            {
+                c = D[c, P[(i + 1) % 8, reversedDigits[i]]];
```
Comment (pradeepgururani) [inline NumberValidator/Validators/IN/vid.cs]: have better names then `D`, `c` and `P`. name the variables in such a manner that they explain their purpose clearly.

File: NumberValidator/Validators/IN/AADHAAR.cs [hunk 8]
```diff
@@ -1,45 +1,47 @@
 ﻿using System;
-using System.Linq;
-using NumberValidator.Helpers;
 using System.Text.RegularExpressions;
 
-namespace NumberValidator.Validators.IN
+namespace NumberValidator.Validators
 {
-    public class Aadhaar : IValidator
+    public interface IValidator
```
Comment (pradeepgururani) [inline NumberValidator/Validators/IN/AADHAAR.cs]: why?

File: NumberValidator/Validators/IN/vid.cs [hunk 9]
```diff
@@ -0,0 +1,149 @@
+﻿using System;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace VidHandling
```
Comment (pradeepgururani) [inline NumberValidator/Validators/IN/vid.cs]: wrong namespace.

**Review by pradeepgururani — CHANGES_REQUESTED** [review]

File: NumberValidator/Validators/IN/AADHAAR.cs [hunk 10]
```diff
@@ -1,45 +1,39 @@
 ﻿using System;
-using System.Linq;
-using NumberValidator.Helpers;
 using System.Text.RegularExpressions;
 
-namespace NumberValidator.Validators.IN
+namespace NumberValidator.Validators
 {
-    public class Aadhaar : IValidator
+    public class AadhaarValidator
```
Comment (pradeepgururani) [inline NumberValidator/Validators/IN/AADHAAR.cs]: every validator needs to implement `IValidator` interface.

File: NumberValidator/Validators/IN/vid.cs [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace NumberValidator
+{
+    public class InvalidLengthException : Exception { }
+    public class InvalidFormatException : Exception { }
+    public class InvalidChecksumException : Exception { }
+
+    public static class Vid
```
Comment (pradeepgururani) [inline NumberValidator/Validators/IN/vid.cs]: implement `IValidator`.

File: NumberValidator/Validators/IN/vid.cs [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    {
+        private static readonly Regex VidRegex = new Regex(@"^[2-9][0-9]{15}$");
+
+        public static string Clear(string number) =>
```
Comment (pradeepgururani) [inline NumberValidator/Validators/IN/vid.cs]: still not using `Clean` method which is provided in the code base.

File: NumberValidator/Validators/IN/vid.cs [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+            for (var index = 0; index < reversedDigits.Length; index++)
+            {
+                checksum = MultiplicationTable[checksum, PermutationTable[(index + 1) % 8, reversedDigits[index]]];
+            }
+
+            return InverseTable[checksum];
+        }
+    }
+
+    class Program
```
Comment (pradeepgururani) [inline NumberValidator/Validators/IN/vid.cs]: what is this? we are using unit testing to test our library implementation. We are developing a reusable library.

File: NumberValidator/Validators/IN/vid.cs (hunk 11 above)
Comment (pradeepgururani) [inline NumberValidator/Validators/IN/vid.cs]: take care of casing for file name of this validator.

File: NumberValidator.Tests/IN/AADHAARFixture.cs [hunk 14]
```diff
@@ -1,27 +1,83 @@
-﻿using Xunit;
-using FluentAssertions;
-using NumberValidator.Helpers;
+﻿using System;
+using Xunit;
+using NumberValidator.Validators;
 
-namespace NumberValidator.Tests.IN
+namespace NumberValidator.Tests
```
Comment (pradeepgururani) [inline NumberValidator.Tests/IN/AADHAARFixture.cs]: bring back `IN`.

File: NumberValidator.Tests/IN/VIDFixture.cs [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+        [Fact]
+        public void Palindrome_ThrowsInvalidFormatException()
+        {
+            Assert.Throws<NumberValidator.InvalidFormatException>(() => Vid.Validate("1234567890123456"));
+        }
+
+        [Fact]
+        public void InvalidPattern_ThrowsInvalidFormatException()
+        {
+            Assert.Throws<NumberValidator.InvalidFormatException>(() => Vid.Validate("0123456789012345"));
```
Comment (pradeepgururani) [inline NumberValidator.Tests/IN/VIDFixture.cs]: remove extra line break.

**Review by pradeepgururani — CHANGES_REQUESTED** [review]

**Review by pradeepgururani — APPROVED** [review]

File: NumberValidator/Validators/IValidator.cs (hunk 1 above)
Comment (AkashSadda) [reply NumberValidator/Validators/IValidator.cs]: its corrected

File: NumberValidator/Validators/IN/vid.cs (hunk 2 above)
Comment (AkashSadda) [reply NumberValidator/Validators/IN/vid.cs]: its corrected

File: NumberValidator/Validators/IN/vid.cs (hunk 3 above)
Comment (AkashSadda) [reply NumberValidator/Validators/IN/vid.cs]: its corrected

File: NumberValidator/Validators/IN/vid.cs (hunk 4 above)
Comment (AkashSadda) [reply NumberValidator/Validators/IN/vid.cs]: its corrected

File: NumberValidator/Validators/IN/vid.cs (hunk 5 above)
Comment (AkashSadda) [reply NumberValidator/Validators/IN/vid.cs]: resolved

File: NumberValidator/Validators/IN/vid.cs (hunk 6 above)
Comment (AkashSadda) [reply NumberValidator/Validators/IN/vid.cs]: resolved

**Pull request merged**